flink-table/flink-table-common/src/test/java/org/apache/flink/table/types/extraction/TypeInferenceExtractorTest.java

```diff
@@ -0,0 +1,679 @@
+				.forScalarFunction(GlobalOutputFunctionHint.class)
+				.expectOutputMapping(
+					InputTypeStrategies.sequence(InputTypeStrategies.explicit(DataTypes.INT())),
+					TypeStrategies.explicit(DataTypes.INT()))
+				.expectOutputMapping(
+					InputTypeStrategies.sequence(InputTypeStrategies.explicit(DataTypes.STRING())),
+					TypeStrategies.explicit(DataTypes.INT())),
+
+			// unsupported output overloading
+			TestSpec
+				.forScalarFunction(InvalidSingleOutputFunctionHint.class)
+				.expectErrorMessage("Function hints that lead to ambiguous results are not allowed."),
+
+			// global and local overloading
+			TestSpec
+				.forScalarFunction(SplitFullFunctionHints.class)
+				.expectOutputMapping(
+					InputTypeStrategies.sequence(InputTypeStrategies.explicit(DataTypes.INT())),
+					TypeStrategies.explicit(DataTypes.INT()))
+				.expectOutputMapping(
+					InputTypeStrategies.sequence(InputTypeStrategies.explicit(DataTypes.BIGINT())),
+					TypeStrategies.explicit(DataTypes.BIGINT())),
+
+			// global and local overloading with unsupported output overloading
+			TestSpec
+				.forScalarFunction(InvalidFullOutputFunctionHint.class)
+				.expectErrorMessage("Function hints with same input definition but different result types are not allowed."),
+
+			// invalid data type hint
+			TestSpec
+				.forScalarFunction(IncompleteFunctionHint.class)
+				.expectErrorMessage("Data type hint does neither specify a data type nor input group for use as function argument."),
+
+			// varargs and ANY input group
+			TestSpec
+				.forScalarFunction(ComplexFunctionHint.class)
+				.expectOutputMapping(
+					InputTypeStrategies.varyingSequence(
+						new String[]{"myInt", "myAny"},
+						new ArgumentTypeStrategy[]{InputTypeStrategies.explicit(DataTypes.INT()), InputTypeStrategies.ANY}),
+					TypeStrategies.explicit(DataTypes.BOOLEAN())),
+
+			// ignore argument names during overloading
+			TestSpec
+				.forScalarFunction(InvalidOutputWithArgNamesFunctionHint.class)
+				.expectErrorMessage("Function hints with same input definition but different result types are not allowed."),
+
+			// global input hints and local output hints
+			TestSpec
+				.forScalarFunction(GlobalInputFunctionHints.class)
+				.expectOutputMapping(
+					InputTypeStrategies.sequence(InputTypeStrategies.explicit(DataTypes.INT())),
+					TypeStrategies.explicit(DataTypes.INT()))
+				.expectOutputMapping(
+					InputTypeStrategies.sequence(InputTypeStrategies.explicit(DataTypes.BIGINT())),
+					TypeStrategies.explicit(DataTypes.INT())),
+
+			// no arguments
+			TestSpec
+				.forScalarFunction(ZeroArgFunction.class)
+				.expectOutputMapping(
+					InputTypeStrategies.sequence(new String[0], new ArgumentTypeStrategy[0]),
+					TypeStrategies.explicit(DataTypes.INT())),
+
+			// test primitive arguments extraction
+			TestSpec
+				.forScalarFunction(MixedArgFunction.class)
+				.expectOutputMapping(
+					InputTypeStrategies.sequence(
+						new String[]{"i", "d"},
```

Comment:
We extract the parameters names via reflection only if we extract the type as well, right? Would it be hard to extract the names also if they were not declared in the FunctionHint?

Reply:
When a `FunctionHint` defines an input, we disable the entire reflective input extraction. This includes the argument names as it would otherwise be pretty complex to make sure that argument names and argument types stay in sync. Argument names are a "nice-to-have" feature, they must not be present therefore, I would suggest to make the logic here not too complex but add a note to the JavaDoc of `FunctionHint#input`.